internal/provider/host_resource.go

```diff
@@ -0,0 +1,554 @@
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/hashicorp/terraform-plugin-framework-jsontypes/jsontypes"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/int64planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ resource.Resource              = &HostResource{}
+	_ resource.ResourceWithConfigure = &HostResource{}
+)
+
+// NewHostResource is a helper function to simplify the provider implementation.
+func NewHostResource() resource.Resource {
+	return &HostResource{}
+}
+
+// HostResource is the resource implementation.
+type HostResource struct {
+	client ProviderHTTPClient
+}
+
+// Metadata returns the resource type name.
+func (r *HostResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_host"
+}
+
+// Configure adds the provider configured client to the resource
+func (r *HostResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*AAPClient)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *AAPClient, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+
+	r.client = client
+}
+
+// Schema defines the schema for the resource.
+func (r *HostResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"inventory_id": schema.Int64Attribute{
+				Required: true,
+			},
+			"instance_id": schema.StringAttribute{
+				Optional: true,
+			},
+			"host_url": schema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"id": schema.Int64Attribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.Int64{
+					int64planmodifier.UseStateForUnknown(),
+				},
+			},
+			"name": schema.StringAttribute{
+				Required: true,
+			},
+			"description": schema.StringAttribute{
+				Optional: true,
+			},
+			"variables": schema.StringAttribute{
+				Optional:   true,
+				CustomType: jsontypes.NormalizedType{},
+			},
+			"enabled": schema.BoolAttribute{
+				Optional:    true,
+				Computed:    true,
+				Description: "Defaults False.",
+			},
+			"groups": schema.ListAttribute{
```

Comment:
I'm getting an error when creating a host and adding it to multiple groups:

```
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to aap_host.foo, provider "provider[\"registry.terraform.io/ansible/aap\"]" produced an unexpected new value:
│ .groups[0]: was cty.NumberIntVal(2), but now cty.NumberIntVal(1).
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to aap_host.foo, provider "provider[\"registry.terraform.io/ansible/aap\"]" produced an unexpected new value:
│ .groups[1]: was cty.NumberIntVal(1), but now cty.NumberIntVal(2).
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
```
I think this may be an ordering issue. In my config I have it as [2, 1], but AAP is showing it as [1, 2]. I'm wondering if we should just make this a set of integers, rather than a list. I don't think we care about the order, and in any case I don't think we can control it.

Reply:
I implemented groups as sets rather than lists. I also had to add a validator https://github.com/ansible/terraform-provider-aap/pull/6/files#diff-44ed19a9988a0367841074b2a279ad45d9f453185f1aa10f5449d2889ceb4025R106
to handle this error https://discuss.hashicorp.com/t/unable-to-handle-empty-list-attribute-in-resource/50952/4 I was also getting.